lib/trie/trie.go

```diff
@@ -333,11 +333,11 @@ func (t *Trie) Put(keyLE, value []byte) {
 
 func (t *Trie) insertKeyLE(keyLE, value []byte, deletedMerkleValues map[string]struct{}) {
 	nibblesKey := codec.KeyLEToNibbles(keyLE)
-	if len(value) == 0 {
-		// Force value to be inserted to nil since we don't
-		// differentiate between nil and empty values.
-		value = nil
-	}
+	// if len(value) == 0 {
+	// 	// Force value to be inserted to nil since we don't
+	// 	// differentiate between nil and empty values.
+	// 	value = nil
+	// }
```

Comment:
Just peeking out of curiosity; we need this and the spec does not differentiate empty byte slices and nil byte slices (it's just 'empty'). The `bitfields and heads must be included every block` error happened before this change as well AFAIK. Finally, one more note is that before #2927 we would have funny trie root hashes (because of empty byte slices and kinda wrong encoding), maybe that could had been the cause?

Reply:
Also, please note that the reason why I have kept this draft is because I am still exploring other solutions.

Reply:
Yes, just peeking out of curiosity as I said.
Try rebasing development and dropping that commit to check if it works fine?

As you said, Rust doesn't have `nil`. For now our trie code only have `nil` node values (and not `{}`), but that can be changed as well to `{}` instead of `nil`, provided we have consistency. Consistency might not even be that needed, but it helps mentally when debugging stuff. What is important is the `n.SubValue != nil` check in

https://github.com/ChainSafe/gossamer/blob/8a17f3779140a736d2e348749b7708963c458558/internal/trie/node/encode.go#L40-L48

You could change that to check for length > 0 instead of a nil check.

Reply:
dropping that commit has a bunch of conflict. So, instead I removed the part which was likely to create such a change, which is this branch. And this branch works fine.